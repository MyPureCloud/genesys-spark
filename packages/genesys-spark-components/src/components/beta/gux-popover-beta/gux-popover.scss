/* This is necessary to override the position of the popover on the page. By default, an HTML popover will
   render in the middle of the screen, which we do not want */
:popover-open {
  position: absolute;
  inset: unset;
}

.gux-popover-wrapper {
  position: absolute;
  inset-block-start: 0;
  inset-inline-start: 0;
  padding: var(--gse-ui-popover-padding);
  overflow: hidden;
  background-color: var(--gse-ui-popover-backgroundColor);
  border: none;
  border-radius: var(--gse-ui-popover-borderRadius);
  box-shadow: var(--gse-ui-popover-boxShadow);

  .gux-arrow {
    position: absolute;
    inline-size: var(--gse-ui-popover-anchor-width);
    block-size: var(--gse-ui-popover-anchor-height);

    /* 4px padding leaves some space for the shadow to bleed into */
    padding-block-end: 4px;
    overflow: hidden;
  }

  /* // TODO: COMUI-3210 - Arrow currently does not show  */
  .gux-arrow-caret {
    inline-size: 0;
    block-size: 0;
    border-block-start: calc(var(--gse-ui-popover-anchor-width) / 2) solid
      var(--gse-ui-popover-backgroundColor);
    border-inline-start: calc(var(--gse-ui-popover-anchor-width) / 2) solid
      transparent;
    border-inline-end: calc(var(--gse-ui-popover-anchor-width) / 2) solid
      transparent;
    filter: drop-shadow(0 0 4px var(--gse-semantic-effects-boxShadow));
  }

  .gux-popover-header {
    display: flex;
    flex-direction: row;
    place-content: center space-between;
    align-items: center;
    padding-block-end: var(--gse-ui-popover-gap);
    font-family: var(--gse-ui-popover-title-text-fontFamily);
    font-size: var(--gse-ui-popover-title-text-fontSize);
    font-weight: var(--gse-ui-popover-title-text-fontWeight);
    line-height: var(--gse-ui-popover-title-text-lineHeight);
    color: var(--gse-ui-popover-headerColor);
  }
}
