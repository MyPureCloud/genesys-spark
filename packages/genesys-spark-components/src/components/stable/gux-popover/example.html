<h1>gux-popover</h1>
<div class="scroll-container">
  <div class="scroll-interior">
    <div id="popover-target">Example Element</div>
    <gux-popover
      id="popover-example"
      position="top"
      for="popover-target"
      is-open
    >
      <span slot="title">Title</span>
      <div class="popover-content-wrapper">
        <div>Popover Content</div>
        <div class="popover-content">
          <gux-button accent="ghost" onclick="notify(event)">
            Action 3
          </gux-button>
          <gux-button accent="secondary" onclick="notify(event)">
            Action 2
          </gux-button>
          <gux-button accent="primary" onclick="notify(event)">
            Action 1
          </gux-button>
        </div>
      </div>
    </gux-popover>
  </div>
</div>

<style
  onload="(function () {
  const popoverTarget = document.getElementById('popover-target');
  const popoverExample = document.getElementById('popover-example');

  popoverTarget.addEventListener('mouseover', () => {
    popoverExample.isOpen = true;
  });
})()"
>
  .scroll-container {
    width: 900px;
    height: 300px;
    margin: auto;
    overflow: scroll;
    border: 1px solid #aaa;
  }

  .scroll-interior {
    position: relative;
    width: 600px;
    height: 600px;
    background-color: #fff;
  }

  .popover-content-wrapper {
    display: flex;
    flex-direction: column;
    gap: 16px;
  }

  .popover-content {
    display: flex;
    flex-direction: row;
    gap: 8px;
    width: 280px;
  }

  #popover-target {
    position: absolute;
    top: 250px;
    left: 250px;
    display: inline-block;
    padding: 10px 15px;
    background-color: #fdfdfd;
    border: 1px #e4e9f0;
    box-shadow: 0 0 2px fade(#202937, 24%);
  }
</style>
