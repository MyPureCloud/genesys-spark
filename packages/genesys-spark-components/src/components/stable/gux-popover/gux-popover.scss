.gux-popover-wrapper {
  position: absolute;
  top: 0;
  left: 0;
  z-index: var(--gse-semantic-zIndex-popover);
  display: inline-block;
  min-width: 280px;
  padding: var(--gse-ui-popover-gap);
  background-color: var(--gse-ui-popover-backgroundColor);
  border-radius: var(--gse-ui-popover-borderRadius);
  box-shadow: var(--gse-ui-popover-boxShadow-x)
    var(--gse-ui-popover-boxShadow-y) var(--gse-ui-popover-boxShadow-blur)
    var(--gse-ui-popover-boxShadow-spread) var(--gse-ui-tooltip-boxShadow-color);

  &.gux-hidden {
    display: none;
  }

  .gux-arrow {
    position: absolute;
    width: var(--gse-ui-popover-anchor-width);
    height: var(--gse-ui-popover-anchor-width);
    background: var(--gse-ui-popover-backgroundColor);
    border-radius: var(--gse-ui-flyoutMenu-arrow-borderRadius);
  }

  .gux-popover-header {
    display: flex;
    flex-direction: row;
    place-content: center space-between;
    align-items: center;
    padding-bottom: var(--gse-ui-popover-gap);
    font-family: var(--gse-ui-popover-title-text-fontFamily);
    font-size: var(--gse-ui-popover-title-text-fontSize);
    font-weight: var(--gse-ui-popover-title-text-fontWeight);
    line-height: var(--gse-ui-popover-title-text-lineHeight);
    color: (--gse-ui-popover-headercolor);
  }

  gux-dismiss-button {
    float: right;
  }
}
