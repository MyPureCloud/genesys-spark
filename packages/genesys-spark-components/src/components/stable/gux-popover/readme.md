# gux-popover

## Possible Improvements

- add a dark theme. There is currently none in the style guide (No Jira)
- add a gux-popover-paragraph component with title and content slots (No Jira)
- add a gux-popover-list component and gux-popover-list-option component (No Jira)
  - It may be possible to use gux-list for this.

<!-- Auto Generated Below -->


## Properties

| Property               | Attribute                | Description                                                                             | Type                                                                                                                                                                 | Default     |
| ---------------------- | ------------------------ | --------------------------------------------------------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ----------- |
| `closeOnClickOutside`  | `close-on-click-outside` | Close popover when the user clicks outside of its bounds                                | `boolean`                                                                                                                                                            | `false`     |
| `displayDismissButton` | `display-dismiss-button` | Indicate if the dismiss button is displayed                                             | `boolean`                                                                                                                                                            | `undefined` |
| `for`                  | `for`                    | Indicates the id of the element the popover should anchor to                            | `string`                                                                                                                                                             | `undefined` |
| `isOpen`               | `is-open`                | Controls hiding and showing the popover                                                 | `boolean`                                                                                                                                                            | `false`     |
| `position`             | `position`               | Indicate position of popover element arrow (follow floating ui placement attribute api) | `"bottom" \| "bottom-end" \| "bottom-start" \| "left" \| "left-end" \| "left-start" \| "right" \| "right-end" \| "right-start" \| "top" \| "top-end" \| "top-start"` | `'bottom'`  |


## Events

| Event        | Description                             | Type                |
| ------------ | --------------------------------------- | ------------------- |
| `guxdismiss` | Fired when a user dismisses the popover | `CustomEvent<void>` |


## Slots

| Slot      | Description            |
| --------- | ---------------------- |
|           | popover content        |
| `"title"` | Slot for popover title |


## Dependencies

### Used by

 - [gux-pagination-ellipsis-button](../gux-pagination/gux-pagination-buttons/gux-pagination-ellipsis-button)
 - [gux-rich-text-editor-action-link](../../beta/gux-rich-text-editor/gux-rich-text-editor-action/gux-rich-text-editor-action-link)

### Depends on

- [gux-dismiss-button](../gux-dismiss-button)

### Graph
```mermaid
graph TD;
  gux-popover --> gux-dismiss-button
  gux-dismiss-button --> gux-button-slot
  gux-dismiss-button --> gux-icon
  gux-pagination-ellipsis-button --> gux-popover
  gux-rich-text-editor-action-link --> gux-popover
  style gux-popover fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
