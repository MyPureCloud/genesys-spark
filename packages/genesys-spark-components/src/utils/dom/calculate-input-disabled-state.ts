export function calculateInputDisabledState(
  element:
    | HTMLButtonElement
    | HTMLFieldSetElement
    | HTMLInputElement
    | HTMLOptGroupElement
    | HTMLOptionElement
    | HTMLSelectElement
    | HTMLTextAreaElement
    | HTMLGuxDropdownElement
    | HTMLGuxDropdownMultiElement
): boolean {
  const fieldSet = element.closest('fieldset');

  return Boolean(element.disabled || fieldSet?.disabled);
}
