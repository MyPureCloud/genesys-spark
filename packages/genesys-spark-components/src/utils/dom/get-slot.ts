export function getSlot(element: HTMLElement, slotName: string): Element {
  // This is a neater solution but our spec test do not support :scope selector
  // return element?.querySelector(`:scope > *[slot=${slotName}]`);

  for (const child of element.children) {
    if (child.matches(`[slot=${slotName}]`)) {
      return child;
    }
  }

  return null;
}
