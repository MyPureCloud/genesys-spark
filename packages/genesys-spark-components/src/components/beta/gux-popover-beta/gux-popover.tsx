import {
  Component,
  Element,
  Event,
  EventEmitter,
  h,
  JSX,
  Listen,
  Prop,
  Watch,
  State
} from '@stencil/core';
import {
  autoUpdate,
  computePosition,
  arrow,
  flip,
  offset,
  Placement,
  shift,
  hide
} from '@floating-ui/dom';

import { OnClickOutside } from '@utils/decorator/on-click-outside';
import { trackComponent } from '@utils/tracking/usage';
import { getSlot } from '@utils/dom/get-slot';
import { findElementById } from '@utils/dom/find-element-by-id';

/**
 * @slot - popover content
 * @slot title - Slot for popover title
 */

@Component({
  styleUrl: 'gux-popover.scss',
  tag: 'gux-popover-beta',
  shadow: true
})
export class GuxPopover {
  private popupElement: HTMLElement;
  private arrowElement: HTMLDivElement;
  private cleanupUpdatePosition: ReturnType<typeof autoUpdate>;

  @Element()
  private root: HTMLElement;

  /**
   * Indicates the id of the element the popover should anchor to
   */
  @Prop()
  for!: string;

  /**
   * Indicate position of popover element arrow (follow floating ui placement attribute api)
   */
  @Prop()
  position: Placement = 'bottom';

  /**
   * Indicate if the dismiss button is displayed
   */
  @Prop()
  displayDismissButton: boolean;

  /**
   * Close popover when the user clicks outside of its bounds
   */
  @Prop()
  closeOnClickOutside: boolean = false;

  /**
   * Controls hiding and showing the popover
   */
  @Prop({ mutable: true })
  isOpen: boolean = false;

  /**
   * Fired when a user dismisses the popover
   */
  @Event()
  guxdismiss: EventEmitter<void>;

  @State()
  private forElement: HTMLInputElement;

  @Watch('for')
  private updateForElement(): void {
    this.forElement = this.getForElement();
    this.forElement.setAttribute('aria-haspopup', 'true');
  }

  @Listen('keydown')
  onKeyDown(event: KeyboardEvent): void {
    switch (event.key) {
      case 'Escape':
        this.forElement.focus();
        break;
    }
  }

  @Listen('focusout')
  onFocusout(event: FocusEvent): void {
    if (!this.closeOnClickOutside && this.displayDismissButton) {
      return;
    }

    const focusIsOutsidePopover =
      event.relatedTarget && !this.root.contains(event.relatedTarget as Node);

    if (focusIsOutsidePopover) {
      this.dismiss();
    }
  }

  @OnClickOutside({ triggerEvents: 'mousedown' })
  checkForClickOutside(event: MouseEvent) {
    const clickPath = event.composedPath();
    const forElement = findElementById(this.root, this.for);
    const clickedForElement = clickPath.includes(forElement);

    if (
      (this.closeOnClickOutside || !this.displayDismissButton) &&
      this.isOpen &&
      !clickedForElement
    ) {
      this.dismiss();
    }
  }

  get titleSlot(): Element | null {
    return getSlot(this.root, 'title');
  }

  private getForElement(): HTMLInputElement {
    if (this.for) {
      const forElement = findElementById(
        this.root,
        this.for
      ) as HTMLInputElement;
      if (!forElement) {
        this.logForAttributeError();
      }

      return forElement;
    } else {
      this.logForAttributeError();
    }
  }

  private focusPopup(): void {
    const autofocusElement: HTMLElement =
      this.root.querySelector('[autoFocus]');
    if (autofocusElement) {
      autofocusElement?.focus();
    } else {
      this.popupElement.focus();
    }
  }

  private runUpdatePosition(): void {
    if (this.root.isConnected) {
      this.cleanupUpdatePosition = autoUpdate(
        findElementById(this.root, this.for),
        this.popupElement,
        () => this.updatePosition(),
        {
          ancestorScroll: true,
          elementResize: true,
          animationFrame: true,
          ancestorResize: true
        }
      );
    } else {
      this.disconnectedCallback();
    }
  }

  private updatePosition(): void {
    const forElement = findElementById(this.root, this.for);

    if (this.popupElement && forElement) {
      void computePosition(forElement, this.popupElement, {
        placement: this.position,
        middleware: [
          offset(7),
          flip(),
          shift(),
          hide(),
          arrow({
            element: this.arrowElement,
            padding: 16
          })
        ]
      }).then(({ x, y, middlewareData, placement }) => {
        Object.assign(this.popupElement.style, {
          left: `${x}px`,
          top: `${y}px`
        });

        const side = placement.split('-')[0];

        const staticSide = {
          top: 'bottom',
          right: 'left',
          bottom: 'top',
          left: 'right'
        }[side];

        const arrowRotation = {
          top: 0,
          right: 90,
          bottom: 180,
          left: -90
        }[side];

        // This is 13 because this makes the arrow look aligned when horizontal
        // or 15 if vertical due to extra padding needed to show the shadow.
        const arrowLen = side === 'left' || side === 'right' ? 15 : 13;

        if (middlewareData.arrow) {
          let x = middlewareData.arrow.x;
          const y = middlewareData.arrow.y;

          if (side === 'left' || side === 'right') {
            x = x + 4;
          }

          this.popupElement.setAttribute('data-placement', placement);

          // TODO: COMUI-3210 - Arrow currently does not show
          Object.assign(this.arrowElement.style, {
            left: x != null ? `${x}px` : '',
            top: y != null ? `${y}px` : '',
            right: '',
            bottom: '',
            [staticSide]: `${-arrowLen}px`,
            transform: `rotate(${arrowRotation}deg)`
          });
        }

        if (middlewareData.hide) {
          Object.assign(this.popupElement.style, {
            visibility: middlewareData.hide.referenceHidden
              ? 'hidden'
              : 'visible'
          });
        }
      });
    }
  }

  private dismiss(): void {
    const dismissEvent = this.guxdismiss.emit();
    if (!dismissEvent.defaultPrevented) {
      this.isOpen = false;
      this.popupElement.togglePopover();
    }
  }

  private logForAttributeError(): void {
    if (this.root.isConnected) {
      console.error(
        `gux-popover: invalid element supplied to 'for': "${this.for}"`
      );
    }
  }

  disconnect: () => void = undefined;

  onKeydown(event: KeyboardEvent): void {
    switch (event.key) {
      case 'Enter':
      case ' ':
        event.preventDefault();
        this.popupElement.togglePopover();
        this.isOpen = !this.isOpen;
        this.runUpdatePosition();
        this.focusPopup();
        break;
    }
  }

  onMouseup(): void {
    this.popupElement.togglePopover();
    this.isOpen = !this.isOpen;
    this.runUpdatePosition();
  }

  connectedCallback(): void {
    this.updateForElement();
    trackComponent(this.root, { variant: this.position });
    const keydownHandler = this.onKeydown.bind(this);
    this.forElement.addEventListener('keydown', keydownHandler);
    const mouseupHandler = this.onMouseup.bind(this);
    this.forElement.addEventListener('mouseup', mouseupHandler);

    this.disconnect = () => {
      this.forElement.removeEventListener('keydown', keydownHandler);
      this.forElement.removeEventListener('mouseup', mouseupHandler);
    };
  }

  componentDidLoad(): void {
    if (!this.forElement) {
      return;
    }

    this.forElement.popoverTargetElement = this.popupElement;

    if (this.isOpen) {
      this.popupElement.togglePopover();
      this.runUpdatePosition();
    }
  }

  componentDidUpdate(): void {
    if (this.isOpen) {
      this.runUpdatePosition();
    } else if (this.cleanupUpdatePosition) {
      this.cleanupUpdatePosition();
    }
  }

  disconnectedCallback(): void {
    if (this.cleanupUpdatePosition) {
      this.cleanupUpdatePosition();
    }
    this.disconnect();
  }

  private renderDismissButton(): JSX.Element {
    if (this.displayDismissButton) {
      return (
        <gux-dismiss-button
          onClick={this.dismiss.bind(this)}
          position="inherit"
        ></gux-dismiss-button>
      ) as JSX.Element;
    }
  }

  render(): JSX.Element {
    return (
      <div
        ref={(el: HTMLDivElement) => (this.popupElement = el)}
        class={{
          'gux-popover-wrapper': true,
          'gux-hidden': !this.isOpen
        }}
        data-placement
        popover="manual"
        tabindex="-1"
      >
        <div
          ref={(el: HTMLDivElement) => (this.arrowElement = el)}
          class="gux-arrow"
        >
          <div class="gux-arrow-caret"></div>
        </div>
        <div class={{ 'gux-popover-header': Boolean(this.titleSlot) }}>
          <slot name="title"></slot>
          {this.renderDismissButton()}
        </div>
        <div class="gux-popover-content">
          <slot />
        </div>
      </div>
    ) as JSX.Element;
  }
}
