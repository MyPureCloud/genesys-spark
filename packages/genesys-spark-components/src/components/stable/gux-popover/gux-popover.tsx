import {
  Component,
  Element,
  Event,
  EventEmitter,
  h,
  JSX,
  Listen,
  Prop
} from '@stencil/core';
import {
  autoUpdate,
  computePosition,
  arrow,
  flip,
  offset,
  Placement,
  shift
} from '@floating-ui/dom';

import { OnClickOutside } from '@utils/decorator/on-click-outside';
import { trackComponent } from '@utils/tracking/usage';
import { getSlot } from '@utils/dom/get-slot';
import { findElementById } from '@utils/dom/find-element-by-id';

/**
 * @slot - popover content
 * @slot title - Slot for popover title
 */

@Component({
  styleUrl: 'gux-popover.scss',
  tag: 'gux-popover',
  shadow: true
})
export class GuxPopover {
  private popupElement: HTMLDivElement;
  private arrowElement: HTMLDivElement;
  private cleanupUpdatePosition: ReturnType<typeof autoUpdate>;

  @Element()
  private root: HTMLElement;

  /**
   * Indicates the id of the element the popover should anchor to
   */
  @Prop()
  for: string;

  /**
   * Indicate position of popover element arrow (follow floating ui placement attribute api)
   */
  @Prop()
  position: Placement = 'bottom';

  /**
   * Indicate if the dismiss button is displayed
   */
  @Prop()
  displayDismissButton: boolean;

  /**
   * Close popover when the user clicks outside of its bounds
   */
  @Prop()
  closeOnClickOutside: boolean = false;

  /**
   * Controls hiding and showing the popover
   */
  @Prop({ mutable: true })
  isOpen: boolean = false;

  /**
   * Fired when a user dismisses the popover
   */
  @Event()
  guxdismiss: EventEmitter<void>;

  @Listen('keydown')
  onKeyDown(event: KeyboardEvent): void {
    switch (event.key) {
      case 'Escape':
        this.dismiss();
        break;
    }
  }

  @Listen('focusout')
  onFocusout(event: FocusEvent): void {
    if (!this.closeOnClickOutside && this.displayDismissButton) {
      return;
    }

    const focusIsOutsidePopover =
      event.relatedTarget && !this.root.contains(event.relatedTarget as Node);

    if (focusIsOutsidePopover) {
      this.dismiss();
    }
  }

  @OnClickOutside({ triggerEvents: 'mousedown' })
  checkForClickOutside(event: MouseEvent) {
    const clickPath = event.composedPath();
    const forElement = findElementById(this.root, this.for);
    const clickedForElement = clickPath.includes(forElement);

    if (
      (this.closeOnClickOutside || !this.displayDismissButton) &&
      this.isOpen &&
      !clickedForElement
    ) {
      this.dismiss();
    }
  }

  get titleSlot(): HTMLSlotElement | null {
    return getSlot(this.root, 'title');
  }

  private runUpdatePosition(): void {
    if (this.root.isConnected) {
      this.cleanupUpdatePosition = autoUpdate(
        findElementById(this.root, this.for),
        this.popupElement,
        () => this.updatePosition(),
        {
          ancestorScroll: true,
          elementResize: true,
          animationFrame: true,
          ancestorResize: true
        }
      );
    } else {
      this.disconnectedCallback();
    }
  }

  private updatePosition(): void {
    const forElement = findElementById(this.root, this.for);

    if (this.popupElement && forElement) {
      void computePosition(forElement, this.popupElement, {
        placement: this.position,
        middleware: [
          offset(7),
          flip(),
          shift(),
          arrow({
            element: this.arrowElement,
            padding: 16
          })
        ]
      }).then(({ x, y, middlewareData, placement }) => {
        Object.assign(this.popupElement.style, {
          left: `${x}px`,
          top: `${y}px`
        });

        const side = placement.split('-')[0];

        const staticSide = {
          top: 'bottom',
          right: 'left',
          bottom: 'top',
          left: 'right'
        }[side];

        const arrowRotation = {
          top: 0,
          right: 90,
          bottom: 180,
          left: -90
        }[side];

        // This is 13 because this makes the arrow look aligned when horizontal
        // or 15 if vertical due to extra padding needed to show the shadow.
        const arrowLen = side === 'left' || side === 'right' ? 15 : 13;

        if (middlewareData.arrow) {
          let x = middlewareData.arrow.x;
          const y = middlewareData.arrow.y;

          if (side === 'left' || side === 'right') {
            x = x + 4;
          }

          this.popupElement.setAttribute('data-placement', placement);

          Object.assign(this.arrowElement.style, {
            left: x != null ? `${x}px` : '',
            top: y != null ? `${y}px` : '',
            right: '',
            bottom: '',
            [staticSide]: `${-arrowLen}px`,
            transform: `rotate(${arrowRotation}deg)`
          });
        }
      });
    }
  }

  private dismiss(): void {
    const dismissEvent = this.guxdismiss.emit();
    if (!dismissEvent.defaultPrevented) {
      this.isOpen = false;
    }
  }

  connectedCallback(): void {
    trackComponent(this.root, { variant: this.position });
  }

  componentDidLoad(): void {
    if (this.isOpen) {
      this.runUpdatePosition();
    }
  }

  componentDidUpdate(): void {
    if (this.isOpen) {
      this.runUpdatePosition();
    } else if (this.cleanupUpdatePosition) {
      this.cleanupUpdatePosition();
    }
  }

  disconnectedCallback(): void {
    if (this.cleanupUpdatePosition) {
      this.cleanupUpdatePosition();
    }
  }

  private renderDismissButton(): JSX.Element {
    if (this.displayDismissButton) {
      return (
        <gux-dismiss-button
          onClick={this.dismiss.bind(this)}
          position="inherit"
        ></gux-dismiss-button>
      ) as JSX.Element;
    }
  }

  render(): JSX.Element {
    return (
      <div
        ref={(el: HTMLDivElement) => (this.popupElement = el)}
        class={{
          'gux-hidden': !this.isOpen,
          'gux-popover-wrapper': true
        }}
        data-placement
      >
        <div
          ref={(el: HTMLDivElement) => (this.arrowElement = el)}
          class="gux-arrow"
        >
          <div class="gux-arrow-caret"></div>
        </div>
        <div class={{ 'gux-popover-header': Boolean(this.titleSlot) }}>
          <slot name="title"></slot>
          {this.renderDismissButton()}
        </div>
        <div class="gux-popover-content">
          <slot />
        </div>
      </div>
    ) as JSX.Element;
  }
}
