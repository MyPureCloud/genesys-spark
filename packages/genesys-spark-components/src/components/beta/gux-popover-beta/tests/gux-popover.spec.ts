jest.mock('@floating-ui/dom', () => {
  const originalModule = jest.requireActual('@floating-ui/dom');

  return {
    __esModule: true,
    ...originalModule,
    autoUpdate: jest.fn()
  } as unknown;
});

import { newSpecPage } from '@test/specTestUtils';
import * as floatingUi from '@floating-ui/dom';
import { MockHTMLElement } from '@stencil/core/mock-doc';

import { GuxPopover } from '../gux-popover';

//Mocks
const showPopover = jest.fn();

const components = [GuxPopover];
const language = 'en';
const html = `
  <div>
    <button id="popover-target">
      Example Element
    </button>
    <gux-popover-beta id="popover-example" position="top" for="popover-target" is-open="false">
      <span slot="title">Title</span>
      <div>popover content</div>
    </gux-popover-beta>
  </div>`;

describe('gux-popover-beta', () => {
  beforeAll(() => {
    Object.assign(MockHTMLElement.prototype, {
      showPopover: showPopover
    });
  });

  afterEach(async () => {
    jest.spyOn(floatingUi, 'autoUpdate').mockRestore();
  });

  describe('#render', () => {
    it('should render popover', async () => {
      jest.spyOn(floatingUi, 'autoUpdate').mockReturnValue(jest.fn());
      jest.spyOn(floatingUi, 'autoUpdate').mockReturnValue(jest.fn());

      const page = await newSpecPage({ components, html, language });

      expect(page.rootInstance).toBeInstanceOf(GuxPopover);
      expect(page.root).toMatchSnapshot();
    });
  });

  describe('#interaction', () => {
    it('should open and close', async () => {
      const cleanupSpy = jest.fn();
      const autoUpdateSpy = jest
        .spyOn(floatingUi, 'autoUpdate')
        .mockReturnValue(cleanupSpy);

      const page = await newSpecPage({ components, html, language });
      const element = page.root as HTMLGuxPopoverElement;
      const targetElement = document.getElementById(
        'popover-target'
      ) as HTMLInputElement;
      const elementWrapper = element.shadowRoot.querySelector(
        '.gux-popover-wrapper'
      );
      targetElement.click();
      element.isOpen = true;
      await page.waitForChanges();
      expect(elementWrapper).not.toHaveClass('gux-hidden');
      expect(autoUpdateSpy).toHaveBeenCalledTimes(1);
      expect(cleanupSpy).not.toBeCalled();

      element.isOpen = false;
      await page.waitForChanges();
      expect(elementWrapper).toHaveClass('gux-hidden');
      expect(cleanupSpy).toHaveBeenCalledTimes(1);

      element.isOpen = true;
      await page.waitForChanges();
      expect(autoUpdateSpy).toHaveBeenCalledTimes(2);
      expect(cleanupSpy).toHaveBeenCalledTimes(2);
    });
  });
});
