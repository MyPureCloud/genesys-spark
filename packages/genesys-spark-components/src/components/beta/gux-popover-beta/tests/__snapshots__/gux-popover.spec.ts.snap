// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gux-popover-beta #render should render popover 1`] = `
<gux-popover-beta for="popover-target" id="popover-example" is-open="false" position="top">
  <template shadowrootmode="open">
    <div class="gux-hidden gux-popover-wrapper" data-placement popover="manual" tabindex="-1">
      <div class="gux-arrow">
        <div class="gux-arrow-caret"></div>
      </div>
      <div class="gux-popover-header">
        <slot name="title"></slot>
      </div>
      <div class="gux-popover-content">
        <slot></slot>
      </div>
      <div>
        <slot name="footer"></slot>
      </div>
    </div>
  </template>
  <span slot="title">
    Title
  </span>
  <div>
    popover content
  </div>
</gux-popover-beta>
`;
